### ISSUE TITLE
about the DR4SR+ traing process

### ISSUE THREAD

**Issue by ComeFromTheMars** [USTC-StarTeam/DR4SR#9]
I need to train a DR4SR+ using SASRec.But this tensor has no gradient during training, so is something wrong?
            **print(meta_loss.grad)**
            **print(meta_train_loss.grad)**
            self.meta_optimizer.step(
                val_loss=**meta_loss**,
                train_loss=**meta_train_loss**,
                aux_params = list(self.meta_module.parameters()),
                parameters = list(self.sub_model.parameters()),
                return_grads = False,
)

**Comment by ShadowTinker** [comment]
It is normal for meta_loss and meta_train_loss in DR4SR+ to have no gradients. This is because, in the bi-level optimization framework, DR4SR+ does not compute gradients through conventional gradient descent methods to update the model. Instead, it directly updates the parameters based on implicit gradient calculations. For more details, please refer to the step code in the MetaOptimizer:

https://github.com/USTC-StarTeam/DR4SR/blob/67aa7ad167dafce081e2a6e788d327cea3b7d83a/utils/utils.py#L235-L240

Here, you will notice that we compute hyper_grads based on meta_loss and meta_train_loss, and then bind these hyper_grads to the parameters that need optimization:

https://github.com/USTC-StarTeam/DR4SR/blob/67aa7ad167dafce081e2a6e788d327cea3b7d83a/utils/utils.py#L242-L243

Finally, through the step operation, the corresponding parameters are updated using the bound hyper_grads (p = p - lr * g):

https://github.com/USTC-StarTeam/DR4SR/blob/67aa7ad167dafce081e2a6e788d327cea3b7d83a/utils/utils.py#L250
